File: python/0114_flatten_binary_tree_to_linked_list.py
# n: node_count
# Time: O(n)
# Space: O(n)
from typing import Optional
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return

        stack = deque()
        if root.right:
            stack.append(root.right)
        if root.left:
            stack.append(root.left)

        prev = root
        while stack:
            curr = stack.pop()
            prev.left = None
            prev.right = curr
            if curr.right:
                stack.append(curr.right)
            if curr.left:
                stack.append(curr.left)
            prev = curr

        return root

File: go/0114_flatten_binary_tree_to_linked_list.go
package main

// n = number of nodes in the tree
// Time:  O(n)
// Space: O(1)
func flatten(root *TreeNode) {
	cur := root

	for cur != nil {
		if cur.Left != nil {
			prev := cur.Left
			for prev.Right != nil {
				prev = prev.Right
			}
			prev.Right = cur.Right
			cur.Right = cur.Left
			cur.Left = nil
		}
		cur = cur.Right
	}
}
